### ISSUE TITLE
Modernising agda2-mode

### ISSUE THREAD

**Issue by phikal** [agda/agda#5917]
The current implementation of `agda2-mode` [appears to assume](https://github.com/agda/agda/blob/5762fccc1b5ffa7c24e35b53c368ebb7c9d827ec/src/data/emacs-mode/agda2-mode.el#L49) it could be used in almost any release of GNU Emacs or XEmacs.  I do not know how rigorously this is tested, but I question the value at the expense of not being able to integrate with modern functionality provided at least from Emacs 25.1+ onward (which is more than 7 years old now).

If there is interest, I could try to rewrite some of the custom code to use common libraries such as Xref or Eldoc.  As implied, this would come at the cost of having to drop XEmacs support and number of old releases.

**Comment by andreasabel** [comment]
Thanks for reaching out, @phikal!
Here is a few base points:
1. Support policy: We generally strive to support an Ubuntu LTS until its end of life.  So we support all the way back to Ubuntu 18.04, which is Emacs 25.2+1 (whatever that means) according to https://repology.org/project/emacs/versions.
2. Testing: We unfortunately lack a framework to test the _Emacs side_ of the interaction facilities of Agda.  (There is a suite in `test/interaction` to test the _Agda side_: Sending request to Agda---that would come from the Emacs agda-mode---and recording Agda's responses---which are in elisp format.)  Thus, changes to the Emacs UX can only be tested manually by some developer, usually just in the Emacs that this developer has currently installed.  The lack of a testing facility makes changes to the agda-mode risky.  This (and lack of elisp skills in many developers) maybe the reason that the agda-mode is rather static (with most of the code 10+ years old).
3. Code owner: We don't follow a _code owner_ policy here, but if you check the [blame](https://github.com/agda/agda/blame/5762fccc1b5ffa7c24e35b53c368ebb7c9d827ec/src/data/emacs-mode/agda2-mode.el#L49), the [original implementation](https://github.com/agda/agda/commit/304ab0d572c46c7af225c6e663ef7e8526346b76#diff-2e7d34756f56895efcfaf99e099987fbe7d1e38414eab56106bd4a6e4cdc393a) of the agda-mode was by Makoto Takeyama (no longer active), and most of the current code is due to Nils Anders Danielsson (@nad).  So I think @nad would be the person to ask and collaborate with concerning refurbishing the agda-mode.

In general, we are in dear need of Elisp-fluent developers in this project.  What would be the scope of your engagement, @phikal?

**Comment by phikal** [comment]
Andreas Abel ***@***.***> writes:

> Thanks for reaching out, @phikal!
> Here is a few base points:
> 1. Support policy: We generally strive to support an Ubuntu LTS until
> its end of life.  So we support all the way back to Ubuntu 18.04,
> which is emac 25.2+1 (whatever that means) according to
> https://repology.org/project/emacs/versions.

That is more than enough.

> 2. Testing: We unfortunately lack a framework to test the _Emacs side_
> of the interaction facilities of Agda.  (There is a suite in
> `test/interaction` to test the _Agda side_: Sending request to
> Agda---that would come from the Emacs agda-mode---and recording Agda's
> responses---which are in elisp format.)  Thus, changes to the Emacs UX
> can only be tested manually by some developer, usually just in the
> Emacs that this developer has currently installed.  The lack of a
> testing facility makes changes to the agda-mode risky.  This (and lack
> of elisp skills in many developers) maybe the reason that the
> agda-mode is rather static (with most of the code 10+ years old).

With more recent versions of Emacs it should be possible to improve the
testing using libraries such as ERT (added with 24.1).

> 3. Code owner: We don't follow a _code owner_ policy here, but if you
> check the
> [blame](https://github.com/agda/agda/blame/5762fccc1b5ffa7c24e35b53c368ebb7c9d827ec/src/data/emacs-mode/agda2-mode.el#L49),
> the [original
> implementation](https://github.com/agda/agda/commit/304ab0d572c46c7af225c6e663ef7e8526346b76#diff-2e7d34756f56895efcfaf99e099987fbe7d1e38414eab56106bd4a6e4cdc393a)
> of the agda-mode was by Makoto Takeyama (no longer active), and most
> of the current code is due to Nils Anders Danielsson ***@***.***).  So I
> think @nad would be the person to ask and collaborate with concerning
> refurbishing the agda-mode.

Is Nils active here?  How should I contact him?

> In general, we are in dear need of Elisp-fluent developers in this
> project.  What would be the scope of your engagement, @phikal?

As I said, I think the code-base can be simplified by using built-in
libraries like Xref, that provide a user interface with a generic
backend that a major mode should concretely implement.  The other
advantage here is that this ensures Emacs behaves more consistently
between various major modes.

From what I have seen up until now, there are also a number of places
where adga2-mode uses a less idiomatic or conventional style that can
also be brushed up.

**Comment by nad** [comment]
I don't think we need to support very old versions of Emacs, and I think it would be great if there was an active maintainer of the Emacs mode.

@phikal, before you do any larger changes I suggest that you discuss them here.

**Comment by phikal** [comment]
Ok, will do.

**Comment by phikal** [comment]
My apologies for the delay in coming back to this issue, here are a few initial proposals:

- Replace `agda2-goto-definition-keyboard' with [Xref](https://www.gnu.org/software/emacs/manual/html_node/emacs/Xref.html). This could include both support for searching variable names/usages as well as Hoogle-like type search.
- Add [Eldoc](https://www.emacswiki.org/emacs/ElDoc) support for type hints in the minibuffer
- Fixing documentation style issues (`M-x checkdoc` currently reports over 50 stylistic problems in `agda2-mode.el`)
- Removing obsolete compatibility support.  There is already an implicit dependency on at least Emacs 24.1, so the concern for XEmacs is pointless.  As mentioned before, I'd recommend raising the minimum version to 25.1.
- Addressing the syntax highlighting situation.  The current approach has difficulties with certain themes, making it difficult to read or recognize some of the things that Agda wishes to point out.  To this end I'd like to investigate if setting `agda2-highlight-face-groups` to `default-faces` is feasible.
- I have been wondering if a more persistent Agda information buffer would be helpful, that would log information sequentially instead of replacing the buffer contents (I see something along these lines has been proposed in #2272)

Otherwise I'd also like to address some of the outstanding issues in the [issue tracker](https://github.com/agda/agda/issues?q=is%3Aopen+is%3Aissue+label%3Aemacs), as well as contribute minor QOL improvements (e.g. using `activate-input-method` instead of `set-input-method` to only update the input method for the current buffer and not the entire Emacs session, and non-Agda related buffers).

**Comment by phikal** [comment]
Depending on how much refactoring is accepted, I believe that addressing the testing situation using [ERT](https://www.gnu.org/software/emacs/manual/html_mono/ert.html) would imaginable.

**Comment by nad** [comment]
@phikal, your list contains a number of different suggestions. Please open separate issues for the (larger) items that you actually plan to work on. For minor things like replacing `set-input-method` with `activate-input-method` it suffices to submit a pull request (the pull request might still be discussed, but creating the pull request should not require a lot of work for you).

**Comment by andreasabel** [comment]
(@phikal: Somehow your comment of 24 May was duplicated in the last comment.)

**Comment by phikal** [comment]
I don't know how that happened, but thank you for the notice.

**Comment by phikal** [comment]
Also see #6548, #6547.

**Comment by VitalyAnkh** [comment]
Hi! @andreasabel @phikal How is this going these days? I'm waiting for these patches to be merged for I can't use agda-mode on Emacs 29 or Emacs master branch due to https://github.com/agda/agda/issues/6525, and this series of patches will resolve the problem. Looking forward to this. :smiling_face_with_tear:

**Comment by phikal** [comment]
Hi, from my perspective everything is ready to be merged. I belive the main issue was waiting for feedback from other Agda users -- the issue is that I don't know if there are no problems, or if nobody has been using it. But you are right, with Emacs 29 coming out soon, at the very least a simple fix for the docstring issue should be pushed at some point.

**Comment by VitalyAnkh** [comment]
Wish at least the docstring issue be fixed soon :rocket: :rocket: :rocket:

**Comment by jespercockx** [comment]
> Hi, from my perspective everything is ready to be merged. I belive the main issue was waiting for feedback from other Agda users -- the issue is that I don't know if there are no problems, or if nobody has been using it. But you are right, with Emacs 29 coming out soon, at the very least a simple fix for the docstring issue should be pushed at some point.

I don't think this will get much more feedback unless it is merged into master, so if it is ready then that's what we should do. The only question is whether it would still be fine to merge this for 2.6.4 or we should wait until after that release (hopefully pretty soon). For now, I will optimistically put this on the 2.6.4 milestone but it might end up getting pushed.

**Comment by jespercockx** [comment]
Since work on this issue seems to have stalled, I am moving it to the `later` milestone. When the issue is resolved, please change it to the current milestone at the time of fixing.

**Comment by phikal** [comment]
I was recently reminded of this issue, and wanted to re-initiate it again.  To make it easier to merge, I would probably go ahead and create separate, much smaller PRs for the important changes, and leave out the cosmetic changes such as removing hanging parentheses.  That means there will be a lot of smaller changes, but I hope that I can propose them independently of one another and increase the chance of them getting through, and not failing like #6536 due to a single commit in a larger PR. Does that sound OK?

**Comment by phikal** [comment]
> Does that sound OK?

Just to respond to my own question, for those who haven't noticed it, I have broken apart the changes that can be considered separately.
